snorkel/classification/training/trainer.py

```diff
@@ -1,27 +1,171 @@
 import logging
 from collections import defaultdict
-from typing import Any, DefaultDict, Dict, List, Optional
+from typing import Any, DefaultDict, Dict, List, Optional, Tuple
 
 import torch
 import torch.nn as nn
 import torch.optim as optim
 from tqdm import tqdm
 
 from snorkel.classification.data import DictDataLoader  # noqa: F401
-from snorkel.classification.snorkel_classifier import SnorkelClassifier
-from snorkel.classification.snorkel_config import default_config
+from snorkel.classification.snorkel_classifier import (
+    ClassifierConfig,
+    SnorkelClassifier,
+)
 from snorkel.classification.training import (
     Checkpointer,
+    CheckpointerConfig,
     LogManager,
+    LogManagerConfig,
     LogWriter,
+    LogWriterConfig,
     TensorBoardWriter,
 )
 from snorkel.classification.training.schedulers import batch_schedulers
-from snorkel.classification.utils import recursive_merge_dicts
+from snorkel.classification.utils import merge_config
+from snorkel.types import Config
 
 Metrics = Dict[str, float]
 
 
+class SGDOptimizerConfig(Config):
+    """Settings for SGD optimizer."""
+
+    momentum: float = 0.9
+
+
+class AdamOptimizerConfig(Config):
+    """Settings for Adam optimizer."""
+
+    amsgrad: bool = False
+    betas: Tuple[float, float] = (0.9, 0.999)
+
+
+class AdamaxOptimizerConfig(Config):
+    """Settings for Adamax optimizer."""
+
+    betas: Tuple[float, float] = (0.9, 0.999)
+    eps: float = 1e-8
+
+
+class OptimizerConfig(Config):
+    """Settings common to all optimizers."""
+
+    lr: float = 0.001
+    l2: float = 0.0
+    grad_clip: float = 1.0
+    sgd_config: SGDOptimizerConfig = SGDOptimizerConfig()  # type:ignore
```

Comment:
this feels a little confusing, given we will only ever have one of these optimizers. 
Can we subclass?

Reply:
or a Union

Reply:
I agree it'd be nice for both the LRScheduler and the Optimizer to not appear to support more than "child" class's parameters, but I think there are issues with both approaches suggested here. When instantiating the AdamOptimizer, for example, the algorithm needs to know what kwargs that optimizer accepts, and unioning the fields would lose that information. And if we subclass OptimizerConfig for the three children, the Trainer still needs to contain all three options so that there's a path to update their values when the Trainer is initialized. I think what we should actually do is just collect "optimizer_kwargs" and "lrscheduler_kwargs" and then pass those to whatever optimizer/scheduler they indicated. Again though, that's orthogonal to this PR of changing the typing of the configs, so I'm going to add that to the tracker for a subsequent update.

Reply:
sg, let's track in a separate PR!